app/models/minion.rb

```diff
@@ -19,27 +19,36 @@ class NonExistingMinion < StandardError; end
   # Example:
   #   Minion.assign_roles(
   #     roles: {
-  #       master: 1,
+  #       master: [1],
   #       minion: [2, 3]
   #     },
   #     default_role: :dns
   #   )
   def self.assign_roles!(roles: {}, default_role: :minion)
-    requested_master = roles[:master].to_i
-    requested_minions =
-      roles[:minion] || Minion.unassigned_role.pluck(:id) - [requested_master]
-    requested_minions = requested_minions.map(&:to_i)
+    requested_masters = select_masters(roles)
+    requested_minions = select_minions(roles, requested_masters)
 
-    if !requested_master.blank? && !Minion.exists?(id: requested_master)
-      raise NonExistingMinion, "Failed to process non existing minion id: #{requested_master}"
+    requested_masters.each do |requested_master|
+      unless Minion.exists?(id: requested_master)
+        raise NonExistingMinion, "Failed to process non existing master id: #{requested_master}"
+      end
+    end
+
+    requested_minions.each do |requested_minion|
+      unless Minion.exists?(id: requested_minion)
+        raise NonExistingMinion, "Failed to process non existing minion id: #{requested_minion}"
+      end
     end
-    master = Minion.find(requested_master)
-    # choose requested minions or all other than master
+
+    # choose selected masters and minions
+    masters = Minion.where(id: requested_masters)
```

Comment:
Again, we have more code duplication: let's merge things into a specific method or something.